# CS4001/4042 Assignment 1, Part B, Q2
In Question B1, we used the Category Embedding model. This creates a feedforward neural network in which the categorical features get learnable embeddings. In this question, we will make use of a library called Pytorch-WideDeep. This library makes it easy to work with multimodal deep-learning problems combining images, text, and tables. We will just be utilizing the deeptabular component of this library through the TabMlp network:

In [1]:
#!pip install pytorch-widedeep

In [2]:
SEED = 42

import os

import random
random.seed(SEED)

import numpy as np
np.random.seed(SEED)

import pandas as pd

from pytorch_widedeep.preprocessing import TabPreprocessor
from pytorch_widedeep.models import TabMlp, WideDeep
from pytorch_widedeep import Trainer
from pytorch_widedeep.metrics import R2Score

>Divide the dataset (‘hdb_price_prediction.csv’) into train and test sets by using entries from the year 2020 and before as training data, and entries from 2021 and after as the test data.

In [3]:
df = pd.read_csv('hdb_price_prediction.csv')

# TODO: Enter your code here

# Training Data
df_train = df[df['year'] <= 2020].copy()
# Testing Data
df_test = df[df['year'] >= 2021].copy()

# Dropping Unncessary Columns
df_train.drop(columns=['year','full_address','nearest_stn'], inplace=True)
df_test.drop(columns=['year','full_address','nearest_stn'], inplace=True)

print("Training Data:", df_train.shape)
print("Testing Data:", df_test.shape)

Training Data: (87370, 11)
Testing Data: (72183, 11)


In [4]:
df.head()

month  year        town              full_address   nearest_stn  \
0      1  2017  ANG MO KIO  406 ANG MO KIO AVENUE 10    Ang Mo Kio   
1      1  2017  ANG MO KIO   108 ANG MO KIO AVENUE 4    Ang Mo Kio   
2      1  2017  ANG MO KIO   602 ANG MO KIO AVENUE 5  Yio Chu Kang   
3      1  2017  ANG MO KIO  465 ANG MO KIO AVENUE 10    Ang Mo Kio   
4      1  2017  ANG MO KIO   601 ANG MO KIO AVENUE 5  Yio Chu Kang   

   dist_to_nearest_stn  dist_to_dhoby  degree_centrality  \
0             1.007264       7.006044           0.016807   
1             1.271389       7.983837           0.016807   
2             1.069743       9.090700           0.016807   
3             0.946890       7.519889           0.016807   
4             1.092551       9.130489           0.016807   

   eigenvector_centrality         flat_model_type  remaining_lease_years  \
0                0.006243        2 ROOM, Improved              61.333333   
1                0.006243  3 ROOM, New Generation              60.583333   
2                0.002459  3 ROOM, New Generation              62.416667   
3                0.006243  3 ROOM, New Generation              62.083333   
4                0.002459  3 ROOM, New Generation              62.416667   

   floor_area_sqm storey_range  resale_price  
0            44.0     10 TO 12      232000.0  
1            67.0     01 TO 03      250000.0  
2            67.0     01 TO 03      262000.0  
3            68.0     04 TO 06      265000.0  
4            67.0     01 TO 03      265000.0

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 159553 entries, 0 to 159552
Data columns (total 14 columns):
 #   Column                  Non-Null Count   Dtype  
---  ------                  --------------   -----  
 0   month                   159553 non-null  int64  
 1   year                    159553 non-null  int64  
 2   town                    159553 non-null  object 
 3   full_address            159553 non-null  object 
 4   nearest_stn             159553 non-null  object 
 5   dist_to_nearest_stn     159553 non-null  float64
 6   dist_to_dhoby           159553 non-null  float64
 7   degree_centrality       159553 non-null  float64
 8   eigenvector_centrality  159553 non-null  float64
 9   flat_model_type         159553 non-null  object 
 10  remaining_lease_years   159553 non-null  float64
 11  floor_area_sqm          159553 non-null  float64
 12  storey_range            159553 non-null  object 
 13  resale_price            159553 non-null  float64
dtypes: float64(7), int64

>Refer to the documentation of Pytorch-WideDeep and perform the following tasks:
https://pytorch-widedeep.readthedocs.io/en/latest/index.html
* Use [**TabPreprocessor**](https://pytorch-widedeep.readthedocs.io/en/latest/examples/01_preprocessors_and_utils.html#2-tabpreprocessor) to create the deeptabular component using the continuous
features and the categorical features. Use this component to transform the training dataset.
* Create the [**TabMlp**](https://pytorch-widedeep.readthedocs.io/en/latest/pytorch-widedeep/model_components.html#pytorch_widedeep.models.tabular.mlp.tab_mlp.TabMlp) model with 2 linear layers in the MLP, with 200 and 100 neurons respectively.
* Create a [**Trainer**](https://pytorch-widedeep.readthedocs.io/en/latest/pytorch-widedeep/trainer.html#pytorch_widedeep.training.Trainer) for the training of the created TabMlp model with the root mean squared error (RMSE) cost function. Train the model for 100 epochs using this trainer, keeping a batch size of 64. (Note: set the *num_workers* parameter to 0.)

In [6]:
cat_embed_cols = [
    ('month', len(np.unique(df['month']))),
    ('town', len(np.unique(df['town']))),
    ('flat_model_type', len(np.unique(df['flat_model_type']))),
    ('storey_range', len(np.unique(df['storey_range']))),
]


continuous_cols = ['dist_to_nearest_stn','dist_to_dhoby','degree_centrality','eigenvector_centrality',
                 'remaining_lease_years','floor_area_sqm']

In [7]:
# TODO: Enter your code here
tab_preprocessor = TabPreprocessor(
    cat_embed_cols = cat_embed_cols, continuous_cols = continuous_cols
)

# Scaled Training Data
X_tab = tab_preprocessor.fit_transform(df_train)

/opt/anaconda3/envs/env_pytorch/lib/python3.9/site-packages/pytorch_widedeep/preprocessing/tab_preprocessor.py:334: UserWarning: Continuous columns will not be normalised
  warnings.warn("Continuous columns will not be normalised")


In [8]:
# Creating the TabMlp Model
model = TabMlp(tab_preprocessor.column_idx, 
                cat_embed_input = tab_preprocessor.cat_embed_input, 
                cat_embed_dropout = 0.1,
                continuous_cols = continuous_cols,
                mlp_hidden_dims = [200, 100])

wide_deep = WideDeep(deeptabular = model) 

In [9]:
# Creating the Trainer 
trainer = Trainer(model = wide_deep,
                  objective = "regression",
                  lr_scheduler_step = False,  
                  num_workers = 0,  
                  metrics = [R2Score])

In [10]:
# Training Model with 100 Epochs with Batch Size 64
trainer.fit(X_tab = X_tab, 
            target = df_train['resale_price'].values, 
            n_epochs = 100, 
            batch_size = 64)

epoch 87: 100%|███| 1366/1366 [00:09<00:00, 144.10it/s, loss=3.59e+9, metrics={'r2': 0.8491}]


epoch 100: 100%|██| 1366/1366 [00:09<00:00, 146.09it/s, loss=3.58e+9, metrics={'r2': 0.8495}]


>Report the test RMSE and the test R2 value that you obtained.

In [13]:
# Scaled Test Data
X_test = tab_preprocessor.transform(df_test)

# Make predictions on the test dataset
y_pred = trainer.predict(X_tab = X_test)

predict: 100%|██████████████████████████████████████████| 1128/1128 [00:02<00:00, 499.53it/s]


In [14]:
y_pred

array([103143.89, 135814.58, 273692.06, ..., 632185.75, 582168.06,
       600626.3 ], dtype=float32)

In [16]:
# TODO: Enter your code here
from sklearn.metrics import r2_score
from sklearn.metrics import mean_squared_error

y_true = df_test['resale_price']

r2 = r2_score(y_true, y_pred)
print("R^2 Score:", r2)
mse = mean_squared_error(y_true, y_pred)
print("Root Mean Squared Error (RMSE):", np.sqrt(mse))

R^2 Score: 0.6162854537452702
Root Mean Squared Error (RMSE): 104798.4517345718


The R^2 value obtained is 0.6162854537452702 and RMSE value is 104798.4517345718